core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/common/accessor/gizmo/GizmoMemberAccessorImplementor.java

```diff
@@ -78,35 +79,55 @@ public String getName() {
-     *        error reporting)
+     * @param className never null
+     * @param classOutput never null, defines how to write the bytecode
+     * @param memberInfo never null, member to generate MemberAccessor methods implementation for
      */
-    public static void defineAccessorFor(ClassCreator classCreator, GizmoMemberDescriptor member,
-            Class<? extends Annotation> annotationClass) {
-        classCreator.getFieldCreator("genericType", Type.class)
-                .setModifiers(Modifier.FINAL);
-        classCreator.getFieldCreator("annotatedElement", AnnotatedElement.class)
-                .setModifiers(Modifier.FINAL);
-
-        GizmoMemberInfo memberInfo = new GizmoMemberInfo(member, annotationClass);
-        // ************************************************************************
-        // MemberAccessor methods
-        // ************************************************************************
-        createConstructor(classCreator, memberInfo);
-        createGetDeclaringClass(classCreator, memberInfo);
-        createGetType(classCreator, memberInfo);
-        createGetGenericType(classCreator);
-        createGetName(classCreator, memberInfo);
-        createGetSpeedNote(classCreator);
-        createSupportSetter(classCreator, memberInfo);
-        createExecuteGetter(classCreator, memberInfo);
-        createExecuteSetter(classCreator, memberInfo);
-        createGetAnnotation(classCreator);
+    public static void defineAccessorFor(String className, ClassOutput classOutput, GizmoMemberInfo memberInfo) {
+        Class<? extends AbstractGizmoMemberAccessor> superClass = getCorrectSuperclass(memberInfo);
+        try (ClassCreator classCreator = ClassCreator.builder()
+                .className(className)
+                .superClass(superClass)
+                .classOutput(classOutput)
+                .setFinal(true)
+                .build()) {
+            classCreator.getFieldCreator("genericType", Type.class)
+                    .setModifiers(Modifier.FINAL);
+            classCreator.getFieldCreator("annotatedElement", AnnotatedElement.class)
+                    .setModifiers(Modifier.FINAL);
+
+            // ************************************************************************
+            // MemberAccessor methods
+            // ************************************************************************
+            createConstructor(classCreator, memberInfo);
+            createGetDeclaringClass(classCreator, memberInfo);
+            createGetType(classCreator, memberInfo);
+            createGetGenericType(classCreator);
+            createGetName(classCreator, memberInfo);
+            createExecuteGetter(classCreator, memberInfo);
+            if (superClass == AbstractReadWriteGizmoMemberAccessor.class) {
+                createExecuteSetter(classCreator, memberInfo);
+            }
+            createGetAnnotation(classCreator);
```

Comment:
Also, can you think of implementing any of these other methods directly, as opposed to generating them?
I did the ones where it was obvious to me, but maybe there are some others where codegen brings no benefit?

Reply:
The only ones I can think of is `getAnnotation`, `getType`, and `getGenericType`, which require the abstract base class to store the AnnotatedElement, return type/field type class, and Type object according. Not so sure how much the Quarkus side would like passing those objects through a recorder though.